software/glasgow/support/bitarray.py

```
@@ -0,0 +1,581 @@
+            yield self.__getbit(bit)
+
+    def __eq__(self, other):
+        try:
+            other = self.__class__(other)
+        except TypeError:
+            return False
+        return self._len_ == other._len_ and self._array_ == other._array_   
+
+    def __iadd__(self, other):
+        other = self.__class__(other)
+        offset = self._len_ % 8
+        self._len_ += other._len_
+        if offset == 0:
+            self._array_.extend(other._array_)
+            return self
+        else:
+            # Merge the adjacent byte of the two buffers first
+            self._array_[-1] &= ~(-1 << offset)
+            self._array_[-1] |= (other._array_[0] << offset) & 0xFF
+            if other._len_ < 8 - offset:
+                return self
+            # Now append the remaining bytes with a bit offset to 
+            # discard the set of already copied bits
+            rem_bits = other._len_ - (8 - offset)
+            rem_bytes = (rem_bits - 1) // 8 + 1
+            prev_array_len = len(self._array_)
+            self._array_.frombytes(bytes(rem_bytes)) # extend buffer
+            self_arr = memoryview(self._array_)      # avoid copies
+            self.__copy_bit_slice(other._array_, 8-offset, 
+                              rem_bits, self_arr[prev_array_len:])
+            return self
+        
+    def __add__(self, other):
+        new = self.__class__(self._array_, self._len_)
+        new += other
+        return new
+
+    def __radd__(self, other):
+        other = self.__class__(other)
+        return other + self
+    
+    def __mul__(self, other):
+        if isinstance(other, int):
+            return reduce(operator.add, (self for _ in range(other-1)), self)
+        return NotImplemented
+
+    def __rmul__(self, other):
+        return self * other
+    
+    def __bitop(self, other, op, clear_top=False):
+        other = self.__class__(other)
+        # Only perform operation in-place when the other bitarray is smaller
```

This seems like a wrong condition to pick between mutating and copying operations.

---

What would be the proper response when `other`'s bit length is larger? Maybe raise an exception like @attie (https://github.com/GlasgowEmbedded/glasgow/pull/265#discussion_r564905060) mentioned? If so, should that only happen for the in-place operations?

---

Changed the behavior in commit ffc277a56af114162cdb7ef050386c01b22b5672